src/entry_stream.rs

```diff
@@ -3,58 +3,421 @@
+}
+
+impl Output for VecOutput {
+    fn write(&mut self, payload: String) -> Result<()> {
+        self.values.push(payload);
+        Ok(())
+    }
+}
+
+impl VecOutput {
+    pub fn new() -> Self {
+        VecOutput { values: Vec::new() }
+    }
+
+    pub fn entries(&self) -> Vec<String> {
+        self.values.clone()
+    }
+}
+
 pub struct MockEntryStream {
-    pub socket: Vec<String>,
+    entry_stream_meta: EntryStreamMeta,
+    output: VecOutput,
 }
 
-impl MockEntryStream {
-    #[allow(clippy::needless_pass_by_value)]
-    pub fn new(_socket: String) -> Self {
-        MockEntryStream { socket: Vec::new() }
+impl std::fmt::Debug for MockEntryStream {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "VecEntryStream {{ last_tick_height: {}, last_slot: {}, last_entry_id: {}, last_leader_id: {} }}",
+            self.entry_stream_meta.last_tick_height,
+            self.entry_stream_meta.last_slot,
+            self.entry_stream_meta.last_entry_id,
+            self.entry_stream_meta.last_leader_id
+        )
     }
 }
 
 impl EntryStreamHandler for MockEntryStream {
-    fn stream_entries(&mut self, entries: &[Entry]) -> Result<()> {
-        for entry in entries {
-            let json = serde_json::to_string(&entry)?;
-            let payload = format!(r#"{{"dt":"{}","entry":{}}}"#, Utc::now().to_rfc3339(), json);
-            self.socket.push(payload);
+    fn emit_entry_events(
+        &mut self,
+        leader_scheduler: &LeaderScheduler,
+        entries: &[Entry],
+    ) -> Result<()> {
+        self.entry_stream_meta
+            .do_entry_events(&mut self.output, leader_scheduler, entries)
+    }
+
+    fn emit_slot_event(
+        &mut self,
+        current_slot: u64,
+        tick_height: u64,
+        leader_id: &Pubkey,
+    ) -> Result<()> {
+        self.entry_stream_meta
+            .do_slot_event(&mut self.output, current_slot, tick_height, leader_id)
+    }
+
+    fn emit_block_event(&mut self) -> Result<()> {
+        self.entry_stream_meta.do_block_event(&mut self.output)
+    }
+}
+
+impl MockEntryStream {
+    pub fn new(_: String) -> Self {
+        MockEntryStream {
+            entry_stream_meta: EntryStreamMeta::new(),
+            output: VecOutput::new(),
         }
-        Ok(())
+    }
+
+    pub fn entries(&self) -> Vec<String> {
+        self.output.entries()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::bank::Bank;
+    use crate::entry::Entry;
+    use crate::leader_scheduler::DEFAULT_TICKS_PER_SLOT;
+    use chrono::{DateTime, FixedOffset};
+    use serde_json::Value;
+    use solana_sdk::hash::Hash;
+
+    #[test]
+    fn test_entry_stream() -> () {
+        // Set up entry stream
+        let mut entry_stream = MockEntryStream::new("test_stream".to_string());
+        let bank = Bank::default();
+        let leader_scheduler = bank.leader_scheduler.read().unwrap();
+        let num_entries = 9;
```

Comment:
This hard-coded number will make this test fail on the tip of master:
https://github.com/solana-labs/solana/blob/56734dca3b8ba0e5f4973d4767ec3299afdef0be/src/leader_scheduler.rs#L24

Reply:
Ok, will update!